### ISSUE TITLE
local variable type refinement limitation

### ISSUE THREAD

**Issue by cushon** [typetools/checker-framework#1751]
The following example is accepted by the nullness checker if the local variable declaration `Optional<String> y` is explicitly annotated as `@Nullable`. Should the checker be able to infer the local variable type?

```java
import java.util.Optional;
import org.checkerframework.checker.nullness.qual.Nullable;

abstract class T {
  interface I<T> {}

  @Nullable abstract <T> T g(I<T> i);

  void f(I<Optional<String>> x) {
    Optional<String> y = g(x);
    if (y != null) {
      System.err.println(y);
    }
  }
}
```

```
./checker-framework-2.3.1/checker/bin/javac -processor Nullness -cp checker-framework-2.3.1/checker/dist/checker-qual.jar T.java
T.java:10: error: [assignment.type.incompatible] incompatible types in assignment.
    Optional<String> y = g(x);
                          ^
  found   : @Initialized @Nullable Optional<@Initialized @NonNull String>
  required: @UnknownInitialization @NonNull Optional<@Initialized @NonNull String>
1 error
```

**Comment by wmdietl** [comment]
Class `java.util.Optional` is by default never allowed to be `@Nullable`, see the `@NonNull` on the `Optional` class declaration https://github.com/typetools/checker-framework/blob/master/checker/jdk/nullness/src/java/util/Optional.java#L63 .
This enforces "Rule #1: Never, ever, use null for an Optional variable or return value." from https://stuartmarks.files.wordpress.com/2016/09/optionalmotherofallbikesheds3.pdf and seems generally accepted practice.
(Explicitly making the local variable `@Nullable` should also be forbidden, as it currently allows to violate that constraint.)

In your example, if you replace `Optional` with `List` throughout, local variable type inference works as expected.

So I would say this is intended behavior and that explicitly writing `@Nullable Optional<...>` should also be forbidden.

**Comment by cushon** [comment]
Thanks for the explanation!

> Never, ever, use null for an Optional variable or return value

I'm aware of that school of thought, but it is not universally accepted. There's a comment about `null` and guava's `Optional` here that still applies: https://stackoverflow.com/questions/9561295/whats-the-point-of-guavas-optional-class#comment12143090_9561334. In our case I think we'd prefer that `Optional` didn't have the `@NonNull` declaration annotation.

But if it does, I agree that disallowing the explicit annotation would have been less confusing.

It might also be helpful for the diagnostic to make it clear that the annotation was coming from the declaration rather than being inferred, but I don't have a specific suggestion for better diagnostic text, and I should have thought to look at the declaration.

**Comment by mernst** [comment]
I agree with both remarks about improving the diagnostics.  If Liam was misled, then I expect almost every user will be.

Though most guidelines agree that Optional should be non-null, we could perhaps give Google a way to disable that particular behavior.

**Comment by wmdietl** [comment]
I think we should remove the `@NonNull` declaration annotation on `Optional` that was introduced in https://github.com/typetools/checker-framework/commit/a9c2179ec2ab3918ac6c6f6e66d2f10ccaf35978#diff-881645d578d9765ea84e0ff165b4cec7R65
`@NonNull` already is the default in most locations. Making `@NonNull` the default for local variables, while still allowing `@Nullable` uses of `Optional`, leads to the inconsistency that Liam observed here.

More generally, I would think that annotations on a class declaration should be an upper bound on all possible uses of that class, not just a default for un-annotated uses. That would have given an error about the `@Nullable Optional` use in this example (the effective return type of `g` in the invocation`), with a more useful error message.

I'm not sure what the best way is to accommodate both people that never want to allow a `@Nullable Optional` and people that want to. Instead of putting this annotation in the annotated JDK, we could put it in a separate stub file to allow restricting `java.util.Optional`, Guava Optional, and any other similar classes, only if desired.

Regarding better diagnostics about where default qualifiers come from: I agree that that would be good. However, we would need to either store with each qualifier where it came from or find a way to reliably re-calculate the types with additional diagnostics.

**Comment by mernst** [comment]
> More generally, I would think that annotations on a class declaration should be an upper bound on all possible uses of that class, not just a default for un-annotated uses.

I agree.
That is already supposed to be the semantics (see https://checkerframework.org/manual/#faq-type-qualifier-on-class-declaration and elsewhere in the manual).  If that isn't the current semantics, then that seems like a bug in the Checker Framework.

**Comment by t-rasmud** [comment]
> More generally, I would think that annotations on a class declaration should be an upper bound on all possible uses of that class, not just a default for un-annotated uses.

The checker-frameworks actually enforces this check in the `BaseTypeVisitor's isValidUse()` method https://github.com/typetools/checker-framework/blob/f93b5ffc571c04a286da2f87cad40fabf18eb3d2/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java#L3470

However, the nullness checker overrides this check with the reasoning given in the comments here

https://github.com/typetools/checker-framework/blob/f93b5ffc571c04a286da2f87cad40fabf18eb3d2/checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java#L146

Is the reasoning incorrect? Should the nullness checker still call the super method?
In general, should all checkers always make a call to the `isValidUse` method in the super class?

**Comment by wmdietl** [comment]
Unfortunately, the comment in `NullnessVisitor` is still correct.
There is no separate category for type declarations and normal defaults apply, making all type declarations `@NonNull`.
If the `isValidUse` implementation were used, all `@Nullable` uses of types would result in an invalid use error!

As outlined in our discussion document, we need to change the default for type declarations to be the top qualifier, which will then allow all uses of that type.
To allow adaptation, we should add `TYPE_DECLARATION` or some such constant to https://github.com/typetools/checker-framework/blob/master/framework/src/main/java/org/checkerframework/framework/qual/TypeUseLocation.java
and then adapt defaulting to apply that qualifier to type declarations.

Once that is done, `NullnessVisitor` can call super.

**Comment by t-rasmud** [comment]
@wmdietl Will this also require changing the default for type declaration in the initialization checker to `@UnknownInitialization`? It is currently set to `@Initialized`.

**Comment by wmdietl** [comment]
Note that the `NullnessVisitor` is a subtype of the `InitializationVisitor`. So once you change `NullnessVisitor.isValidUse` to invoke `super`, the Initialization hierarchy probably will run into the same issue.
So yes, I think the default for type declarations for the Initialization Checker should also be the top type, in this case `@UnknownInitialization`.

**Comment by wmdietl** [comment]
#2247 is the dedicated bug for the remaining tasks.
